### PR TITLE
stake-contract: remove phoenix-core dependency

### PR DESCRIPTION
**Pull request by miloszm**
Removes phoenix-core dependency in stake contract, implements #1138

### CODE REVIEW

File: contracts/stake-types/src/sig.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+const ALLOW_MESSAGE_SIZE: usize = u64::SIZE + PublicKey::SIZE;
+const STAKE_MESSAGE_SIZE: usize = u64::SIZE + u64::SIZE;
+const WITHDRAW_MESSAGE_SIZE: usize =
+    u64::SIZE + StealthAddress::SIZE + BlsScalar::SIZE;
+
+/// Signature message used for [`Allow`].
+#[must_use]
+pub fn allow_signature_message(
+    counter: u64,
+    staker: PublicKey,
```
Comment (herr-seppia) [inline contracts/stake-types/src/sig.rs:25]: Can you take a reference here?

File: contracts/transfer-types/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#[derive(Debug, Clone, Archive, Deserialize, Serialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct Mint {
+    /// The address to mint to.
+    pub address: StealthAddress,
+    /// The value to mint to the address.
+    pub value: u64,
+    /// A nonce to prevent replay.
+    pub nonce: BlsScalar,
+}
```
Comment (herr-seppia) [inline contracts/transfer-types/src/lib.rs:59]: I believe this is not related to this PR. Instead of this, can you migrate the whole icc messages? (wfct, stct, wfco, stco)

Comment (miloszm) [reply contracts/transfer-types/src/lib.rs:59]: Agree with adding `wfco` and `stco`, yet Mint is needed for this PR, it is used by stake contract.

Comment (herr-seppia) [reply contracts/transfer-types/src/lib.rs:134]: Ok, then can you migrate TreeLeaf too? So we can get rid of `phoenix_core::transaction::transfer` completely

**Review by herr-seppia — CHANGES_REQUESTED** [review]
LGTM overall

pls add the TreeLeaf struct too and rebase

**Review by herr-seppia — APPROVED** [review]
LGTM
